barretenberg/cpp/src/barretenberg/vm2/tracegen/sha256_trace.cpp

```diff
@@ -0,0 +1,355 @@
+    0xe49b69c1, 0xefbe4786, 0x0fc19dc6, 0x240ca1cc, 0x2de92c6f, 0x4a7484aa, 0x5cb0a9dc, 0x76f988da,
+    0x983e5152, 0xa831c66d, 0xb00327c8, 0xbf597fc7, 0xc6e00bf3, 0xd5a79147, 0x06ca6351, 0x14292967,
+    0x27b70a85, 0x2e1b2138, 0x4d2c6dfc, 0x53380d13, 0x650a7354, 0x766a0abb, 0x81c2c92e, 0x92722c85,
+    0xa2bfe8a1, 0xa81a664b, 0xc24b8b70, 0xc76c51a3, 0xd192e819, 0xd6990624, 0xf40e3585, 0x106aa070,
+    0x19a4c116, 0x1e376c08, 0x2748774c, 0x34b0bcb5, 0x391c0cb3, 0x4ed8aa4a, 0x5b9cca4f, 0x682e6ff3,
+    0x748f82ee, 0x78a5636f, 0x84c87814, 0x8cc70208, 0x90befffa, 0xa4506ceb, 0xbef9a3f7, 0xc67178f2
+};
+
+}; // namespace
+
+// These are helper functions to iterate and set repetitive columns in the trace.
+void Sha256TraceBuilder::set_helper_cols(std::array<uint32_t, 16> prev_w_helpers, TraceContainer& trace)
+{
+    for (size_t i = 0; i < 16; i++) {
+        trace.set(row, { { { w_cols[i], prev_w_helpers[i] } } });
+    }
+}
+
+void Sha256TraceBuilder::set_state_cols(std::array<uint32_t, 8> state, TraceContainer& trace)
+{
+    for (size_t i = 0; i < 8; i++) {
+        trace.set(row, { { { state_cols[i], state[i] } } });
+    }
+}
+
+void Sha256TraceBuilder::set_init_state_cols(std::array<uint32_t, 8> init_state, TraceContainer& trace)
+{
+    for (size_t i = 0; i < 8; i++) {
+        trace.set(row, { { { init_state_cols[i], init_state[i] } } });
+    }
+}
+
+// Decomposes a into two 32-bit values at the bit position b and inserts witness data into the trace.
+void Sha256TraceBuilder::into_limbs_with_witness(
+    uint64_t a, const uint8_t b, TraceContainer& trace, Column c_lhs, Column c_rhs)
+{
+    uint32_t a_lhs = static_cast<uint32_t>(a >> b);
+    uint32_t a_rhs = static_cast<uint32_t>(a) & static_cast<uint32_t>((static_cast<uint64_t>(1) << b) - 1);
+    trace.set(row, { { { c_lhs, a_lhs }, { c_rhs, a_rhs } } });
+}
+
+// Performs 32-bit rotation with witness data inserted into the trace.
+uint32_t Sha256TraceBuilder::ror_with_witness(
+    const uint32_t val, const uint8_t shift, TraceContainer& trace, Column c_result, Column c_lhs, Column c_rhs)
+{
+    auto result = (val >> (shift & 31U)) | (val << (32U - (shift & 31U)));
+    into_limbs_with_witness(val, shift, trace, c_lhs, c_rhs);
+    trace.set(c_result, row, result);
+    return result;
+}
+
+// Performs 32-bit shift right with witness data inserted into the trace.
+uint32_t Sha256TraceBuilder::shr_with_witness(
+    const uint32_t val, const uint8_t shift, TraceContainer& trace, Column c_result, Column c_lhs, Column c_rhs)
+{
+    auto result = val >> shift;
+    into_limbs_with_witness(val, shift, trace, c_lhs, c_rhs);
+    trace.set(c_result, row, result);
+    return result;
+}
+
+// Computes and returns the message schedule (w) value for that round, and inserts witness data into the trace.
+uint32_t Sha256TraceBuilder::compute_w_with_witness(std::array<uint32_t, 16> prev_w_helpers, TraceContainer& trace)
```

Comment:
I will consider taking the trace in the constructor so that we can eventually avoid passing it everywhere.